mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -328,6 +330,38 @@ class WidgetAdapter(
         }
     }
 
+    abstract class HeavyDataViewHolder internal constructor(
+        inflater: LayoutInflater,
+        parent: ViewGroup,
+        @LayoutRes layoutResId: Int
+    ) : ViewHolder(inflater, parent, layoutResId) {
+        protected val widgetContentView: View = itemView.findViewById(R.id.widget_content)
+        private val dataSaverView: View = itemView.findViewById(R.id.data_saver)
+        private val dataSaverButton: Button = itemView.findViewById(R.id.data_saver_button)
+        private val dataSaverHint: TextView = itemView.findViewById(R.id.data_saver_hint)
+        private var overrideDataSaver: Boolean = false
```

Comment:
Having a plain member variable for this doesn't work (as intended). ViewHolders can be re-bound to different widgets at any time, which needs special handling. IMHO, the flow should be something like
- on rebind, check whether the request can be satisfied easily (e.g. image is already cached)
- if that isn't the case, set flag to false, show button etc.
- if it is the case, just show the image

(To test this, just try having e.g. 10 image items with alternating URLs -> item 1 with URL 1, item 2 with URL 2, item 3 with URL 1, etc.)

Reply:
I thought about having a global list with ids of whitelisted widgets. It's totally fine to save this list in ram only, so the after the app is stopped, all widgets with heavy data usage are removed from the whitelist.

Reply:
> on rebind, check whether the request can be satisfied easily

How can I do that for videos? They aren't cached.

Reply:
You probably can't. Just unset the flag in that case.

Reply:
This is why I prefer a different way to whitelist widgets. Can you have a look at https://github.com/openhab/openhab-android/pull/1796/commits/7a13ecf31a36d60bce11ddaac1f4f057d4b688f9?

Reply:
I'm not sure this approach is meaningful. Assume data saver is on, and you chose to play a video nevertheless. If you scroll down to another widget and scroll up again, do you really want to auto-load the video again, using up your data for a *second* time?

Reply:
Fair enough. I implemented your suggestion: https://github.com/openhab/openhab-android/pull/1796/commits/eceea36e1194124286b80c6de38c4c1f383c9d91
However there's a side effect: Swipe to refresh now clears the data saver state of any image.